src/Middleware/RecordTransaction.php

```diff
@@ -64,12 +65,48 @@ public function handle($request, Closure $next)
             'type' => 'HTTP'
         ]);
 
-        $transaction->setSpans(app('query-log')->toArray());
+        foreach (app('apm-spans-log')->toArray() as $spanContext) {
+            // @see https://www.elastic.co/guide/en/apm/server/master/exported-fields-apm-span.html
+            $spanDb = new Span($spanContext['name'], $transaction);
+            $spanDb->setType($spanContext['type']);
+            $spanDb->setSubtype($spanContext['subtype']);
+            $spanDb->setContext($spanContext['context']);
+
+            // optiponal fields
+            if (isset($spanContext['action'])) {
+                $spanDb->setAction($spanContext['action']);
+            }
+            if (isset($spanContext['stacktrace'])) {
+                $spanDb->setStacktrace($spanContext['stacktrace']->toArray());
+            }
+
+            $spanDb->start();
+            $spanDb->stop($spanContext['duration']); // in [ms]
+
+            $spanDb->setStart($spanContext['start']); // in [us]
+
+            $this->agent->putEvent($spanDb);
+        }
 
         if (config('elastic-apm.transactions.use_route_uri')) {
-            $transaction->setTransactionName($this->getRouteUriTransactionName($request));
+            if (config('elastic-apm.transactions.normalize_uri')) {
+                $transaction->setTransactionName($this->getMormalizedTransactionName($request));
+            }
+            else {
+                $transaction->setTransactionName($this->getRouteUriTransactionName($request));
+            }
         }
 
+        // handle X-Requested-By header
+        $requestedBy = $request->headers->get('X-Requested-By', 'end-user');
+
+        // X-Requested-With: XMLHttpRequest (AJAX requests)
+        if ($request->headers->get('X-Requested-With') === 'XMLHttpRequest') {
+            $requestedBy = 'end-user-ajax';
```

Comment:
Is it possible that `$request->headers->get('X-Requested-With') === 'XMLHttpRequest'` but it's not end user that sent the request? I don't think so, but if it's possible then we would override the correct header here.

Reply:
I don't think that any backend service would ever set this header. Unless someone comes up with an idea of server-site AJAX requests ;)

Reply:
https://github.com/bethinkpl/echo-server/blob/134f9093392cc098660a0e945999811aad983500/src/channels/private-channel.ts#L140 - :man_facepalming:

Reply:
bba796a